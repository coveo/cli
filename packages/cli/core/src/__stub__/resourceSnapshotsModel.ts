import {
  ResourceSnapshotsModel,
  ResourceSnapshotsReportModel,
} from '@coveo/platform-client';

export const getDummySnapshotModel = (
  orgId: string,
  snapshotId: string,
  reports: ResourceSnapshotsReportModel[] = []
): ResourceSnapshotsModel => ({
  id: snapshotId,
  createdBy: 'user@coveo.com',
  createdDate: 1622555047116,
  targetId: orgId,
  developerNote: 'hello',
  reports: reports,
  synchronizationReports: [],
  contentSummary: {EXTENSION: 1, FIELD: 11},
});
