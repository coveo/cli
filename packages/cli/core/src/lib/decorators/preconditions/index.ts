// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
