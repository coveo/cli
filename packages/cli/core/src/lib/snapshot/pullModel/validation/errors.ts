import {ValidationError} from 'jsonschema';
import {CLIBaseError} from '@coveo/cli-commons/errors/cliBaseError';
import {getPrettyJsonValidationErrors} from '@coveo/cli-commons/errors/jsonError';

abstract class BaseSPMError extends CLIBaseError {
  public name = 'Snapshot Pull Model Error';
  private static messageSuffix =
    '\nThis is probably a problem with the Coveo CLI, please report this issue at https://github.com/coveo/cli/issues';
  public constructor(message: string, shouldContactCoveo: boolean) {
    super(`${message}${shouldContactCoveo ? BaseSPMError.messageSuffix : ''}`);
  }
}

export class UnknownSPMValidationError extends BaseSPMError {
  public name = 'Unknown Snapshot Pull Model Error';
  public constructor(shouldContactCoveo: boolean) {
    super(
      "An unknown error occured while validating the custom template. Try recreating it from 'empty' or 'full'.",
      shouldContactCoveo
    );
  }
}

export class InvalidSPMError extends BaseSPMError {
  public name = 'Invalid Snapshot Pull Model Error';
  public constructor(shouldContactCoveo: boolean, errors: ValidationError[]) {
    super(getPrettyJsonValidationErrors(errors), shouldContactCoveo);
  }
}
