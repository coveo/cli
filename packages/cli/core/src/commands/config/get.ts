import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
