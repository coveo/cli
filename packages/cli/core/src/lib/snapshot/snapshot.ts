import {
  ResourceSnapshotsModel,
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportResultCode,
  ResourceSnapshotsReportStatus,
  PlatformClient,
  ResourceSnapshotsReportType,
  SnapshotExportContentFormat,
  ApplyOptionsDeletionScope,
} from '@coveo/platform-client';
import retry from 'async-retry';
import {ReportViewer} from './reportPreviewer/reportPreviewer';
import {ensureFileSync, writeJsonSync} from 'fs-extra';
import {join} from 'path';
import {SnapshotReporter} from './snapshotReporter';
import {SnapshotOperationTimeoutError} from '../errors';
import {ExpandedPreviewer} from './expandedPreviewer/expandedPreviewer';
import {Project} from '../project/project';
import {SnapshotNoReportFoundError} from '../errors/snapshotErrors';
import {SnapshotReportStatus} from './reportPreviewer/reportPreviewerDataModels';

export type SnapshotReport = ResourceSnapshotsReportModel;

export interface WaitUntilDoneOptions {
  /**
   * The maximum number of seconds to wait before the commands exits with a timeout error.
   * A value of zero will prevent the command from timing out.
   */
  wait?: number; // in seconds
  /**
   * The interval between 2 consecutive polls.
   */
  waitInterval?: number; // in seconds
  /**
   * Callback to execute every time a request is being made to retrieve the snapshot data
   */
  onRetryCb?: (report: ResourceSnapshotsReportModel) => void;
}

export interface WaitUntilOperationDone extends WaitUntilDoneOptions {
  /**
   * The operation to wait for. If not specified, the method will wait for any operation to complete.
   */
  operationToWaitFor?: ResourceSnapshotsReportType;
}

export class Snapshot {
  public static defaultWaitOptions: Required<WaitUntilDoneOptions> = {
    waitInterval: 1,
    wait: 60,
    onRetryCb: (_report: ResourceSnapshotsReportModel) => {},
  };

  private static ongoingReportStatuses = [
    ResourceSnapshotsReportStatus.Pending,
    ResourceSnapshotsReportStatus.InProgress,
  ];

  public constructor(
    private model: ResourceSnapshotsModel,
    private client: PlatformClient
  ) {}

  public async validate(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ): Promise<SnapshotReporter> {
    await this.snapshotClient.dryRun(this.id, {
      deleteMissingResources,
    });

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.DryRun,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async preview(
    projectToPreview: Project,
    deleteMissingResources = false,
    expandedPreview = true
  ) {
    const reporter = new SnapshotReporter(this.latestReport);
    await this.displayLightPreview(reporter);
    if (!expandedPreview) {
      return;
    }
    const onSuccess = () =>
      this.generateExpandedPreview(projectToPreview, deleteMissingResources);
    await reporter
      .setReportHandler(
        SnapshotReportStatus.SUCCESS,
        async function (this: SnapshotReporter) {
          await onSuccess();
        }
      )
      .handleReport();
  }

  public async apply(
    deleteMissingResources = false,
    options: WaitUntilDoneOptions = {}
  ) {
    await this.snapshotClient.apply(
      this.id,
      deleteMissingResources
        ? {
            deleteMissingResources,
            deletionScope: ApplyOptionsDeletionScope.OnlyTypesFromSnapshot,
          }
        : {deleteMissingResources}
    );

    await this.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.Apply,
      ...options,
    });

    return new SnapshotReporter(this.latestReport);
  }

  public async delete() {
    await this.client.resourceSnapshot.delete(this.model.id);
  }

  public download() {
    return this.client.resourceSnapshot.export(this.id, {
      contentFormat: SnapshotExportContentFormat.SplitPerType,
    });
  }

  public areResourcesInError() {
    return (
      this.latestReport.resultCode ===
      ResourceSnapshotsReportResultCode.ResourcesInError
    );
  }

  public saveDetailedReport(projectPath: string) {
    const pathToReport = join(
      projectPath,
      'snapshot-reports',
      `${this.latestReport.id}.json`
    );
    ensureFileSync(pathToReport);
    writeJsonSync(pathToReport, this.latestReport, {spaces: 2});
    return pathToReport;
  }

  public get latestReport() {
    const reports = this.model.reports;
    if (!Array.isArray(reports) || reports.length === 0) {
      throw new SnapshotNoReportFoundError(this);
    }
    return this.sortReportsByDate(reports)[0];
  }

  public get id() {
    return this.model.id;
  }

  public get targetId() {
    return this.model.targetId;
  }

  private get snapshotClient() {
    return this.client.resourceSnapshot;
  }

  private sortReportsByDate<T extends SnapshotReport>(report: T[]): T[] {
    return report.sort((a, b) => b.updatedDate - a.updatedDate);
  }

  private async displayLightPreview(reporter: SnapshotReporter) {
    const viewer = new ReportViewer(reporter);
    await viewer.display();
  }

  private async generateExpandedPreview(
    projectToPreview: Project,
    shouldDelete: boolean
  ) {
    const previewer = new ExpandedPreviewer(
      this.latestReport,
      this.targetId,
      projectToPreview,
      shouldDelete
    );
    await previewer.preview();
  }

  private async refreshSnapshotData() {
    this.model = await this.snapshotClient.get(this.model.id, {
      includeReports: true,
    });
  }

  public waitUntilDone(options: WaitUntilOperationDone = {}) {
    const opts = {...Snapshot.defaultWaitOptions, ...options};
    const toMilliseconds = (seconds: number) => seconds * 1e3;

    return retry(
      this.waitUntilDoneRetryFunction(opts.onRetryCb, opts.operationToWaitFor),
      // Setting the retry mechanism to follow a time-based logic instead of specifying the  number of attempts.
      {
        retries: Math.ceil(opts.wait / opts.waitInterval),
        forever: opts.wait === 0,
        minTimeout: toMilliseconds(opts.waitInterval),
        maxTimeout: toMilliseconds(opts.waitInterval),
        maxRetryTime: toMilliseconds(opts.wait),
      }
    );
  }

  private isUnsettled() {
    return Snapshot.ongoingReportStatuses.includes(this.latestReport.status);
  }

  private isGoingThroughOperation(
    operation: ResourceSnapshotsReportType
  ): boolean {
    if (this.latestReport.type === operation) {
      return true;
    }
    return false;
  }

  private waitUntilDoneRetryFunction(
    onRetryCb: (report: ResourceSnapshotsReportModel) => void,
    operationToWaitFor?: ResourceSnapshotsReportType
  ): () => Promise<void> {
    return (async () => {
      await this.refreshSnapshotData();

      const isUnsettled = this.isUnsettled();
      const isUnexpectedOperation = operationToWaitFor
        ? !this.isGoingThroughOperation(operationToWaitFor)
        : false;

      onRetryCb(this.latestReport);

      if (isUnsettled || isUnexpectedOperation) {
        throw new SnapshotOperationTimeoutError(this);
      }
    }).bind(this);
  }
}
