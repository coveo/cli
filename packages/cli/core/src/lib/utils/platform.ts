import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
