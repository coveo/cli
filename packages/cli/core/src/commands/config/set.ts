import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
