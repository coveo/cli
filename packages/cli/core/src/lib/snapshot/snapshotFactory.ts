import {
  CreateFromFileOptions,
  ResourceSnapshotsReportType,
  ResourceSnapshotType,
} from '@coveo/platform-client';
import {readFileSync} from 'fs';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {SnapshotPullModelResources} from './pullModel/interfaces';
import {Snapshot, WaitUntilDoneOptions} from './snapshot';
import {Project} from '../project/project';
import {ensureResourcesAccess, ensureSnapshotAccess} from './snapshotAccess';
import {Blob} from 'node:buffer';

export class SnapshotFactory {
  public static async createSnapshotFromProject(
    project: Project,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ): Promise<Snapshot> {
    const client = await this.getClient(targetOrg);
    await ensureResourcesAccess(client, project.resourceTypes);
    const pathToZip = await project.compressResources();
    const file = readFileSync(pathToZip);
    const blob = new Blob([file], {
      type: 'application/zip',
    });
    const computedOptions: CreateFromFileOptions = {
      developerNotes: 'cli-created-from-zip',
    };

    const model = await client.resourceSnapshot.createFromFile(
      blob,
      computedOptions
    );
    const snapshot = new Snapshot(model, client);

    await snapshot.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.CreateSnapshot,
      ...options,
    });

    return snapshot;
  }

  public static async createFromExistingSnapshot(
    snapshotId: string,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ) {
    const client = await this.getClient(targetOrg);
    await ensureSnapshotAccess(client, snapshotId);
    const model = await client.resourceSnapshot.get(snapshotId, {
      includeReports: true,
    });

    const snapshot = new Snapshot(model, client);
    await snapshot.waitUntilDone(options);

    return snapshot;
  }

  public static async createFromOrg(
    resourcesToExport: SnapshotPullModelResources,
    targetOrg: string,
    options?: WaitUntilDoneOptions
  ) {
    const client = await this.getClient(targetOrg);
    const resourceTypes = Object.keys(
      resourcesToExport
    ) as ResourceSnapshotType[];
    await ensureResourcesAccess(client, resourceTypes);
    const model = await client.resourceSnapshot.createFromOrganization(
      {resourcesToExport},
      {includeChildrenResources: true, developerNotes: 'Created by Coveo-CLI'}
    );

    const snapshot = new Snapshot(model, client);

    await snapshot.waitUntilDone({
      operationToWaitFor: ResourceSnapshotsReportType.CreateSnapshot,
      ...options,
    });

    return snapshot;
  }

  private static async getClient(targetOrg: string) {
    return await new AuthenticatedClient().getClient({organization: targetOrg});
  }
}
