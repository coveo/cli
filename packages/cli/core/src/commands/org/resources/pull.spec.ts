jest.mock('../../../lib/decorators/preconditions/git');
jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/platform/authenticatedClient');
jest.mock('../../../lib/snapshot/snapshotFactory');
jest.mock('../../../lib/project/project');
jest.mock('../../../lib/utils/process');

import {join} from 'path';
import {Config} from '@coveo/cli-commons/config/config';
import {
  ResourceSnapshotsReportType,
  ResourceSnapshotType,
  SnapshotAccessType,
} from '@coveo/platform-client';
import {test} from '@oclif/test';
import {getDummySnapshotModel} from '../../../__stub__/resourceSnapshotsModel';
import {getSuccessReport} from '../../../__stub__/resourceSnapshotsReportModel';
import {SnapshotFactory} from '../../../lib/snapshot/snapshotFactory';
import {Snapshot} from '../../../lib/snapshot/snapshot';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {Command, CliUx} from '@oclif/core';
import {PreconditionError} from '@coveo/cli-commons/errors/preconditionError';
import {cwd} from 'process';
import {Project} from '../../../lib/project/project';
import {IsGitInstalled} from '../../../lib/decorators/preconditions';
import {
  MissingResourcePrivileges,
  MissingSnapshotPrivilege,
} from '../../../lib/errors/snapshotErrors';

const mockedSnapshotFactory = jest.mocked(SnapshotFactory);
const mockedProject = jest.mocked(Project);
const mockedConfig = jest.mocked(Config);
const mockedConfigGet = jest.fn();
const mockedGetSnapshot = jest.fn();
const mockedDownloadSnapshot = jest.fn();
const mockedDeleteSnapshot = jest.fn();
const mockedIsGitInstalled = jest.mocked(IsGitInstalled);
const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
const mockEvaluate = jest.fn();
const pathToStub = join(cwd(), 'src', '__stub__');

const doMockConfig = () => {
  mockedConfigGet.mockReturnValue({
    region: 'us',
    organization: 'default-org',
    environment: 'prod',
  });

  mockedConfig.mockImplementation(
    () =>
      ({
        get: mockedConfigGet,
      } as unknown as Config)
  );
};

const doMockPreconditions = () => {
  const preconditionStatus = {
    git: true,
  };

  const thrower = () => {
    throw new PreconditionError('Precondition error');
  };
  const mockGit = function (_target: Command) {
    return new Promise<void>((resolve) =>
      preconditionStatus.git ? resolve() : thrower()
    );
  };
  mockedIsGitInstalled.mockReturnValue(mockGit);
};

const mockAuthenticatedClient = () => {
  const mockGetClient = jest.fn().mockResolvedValue({
    privilegeEvaluator: {
      evaluate: mockEvaluate,
    },
  });

  mockedAuthenticatedClient.prototype.getClient = mockGetClient;
};

const mockUserHavingAllRequiredPlatformPrivileges = () => {
  mockEvaluate.mockResolvedValue({approved: true});
};

const mockUserNotHavingAllRequiredPlatformPrivileges = () => {
  mockEvaluate.mockResolvedValue({approved: false});
};

const mockUserNotHavingAllRequiredResourcePrivileges = () => {
  mockedSnapshotFactory.createFromOrg.mockImplementation(() => {
    throw new MissingResourcePrivileges(
      [ResourceSnapshotType.source, ResourceSnapshotType.mlModel],
      SnapshotAccessType.Read
    );
  });
  mockedSnapshotFactory.createFromExistingSnapshot.mockImplementation(() => {
    throw new MissingSnapshotPrivilege(
      'some-snapshot-id',
      SnapshotAccessType.Read
    );
  });
};

const doMockSnapshotFactory = () => {
  mockedSnapshotFactory.createFromOrg.mockResolvedValue({
    delete: mockedDeleteSnapshot,
    download: mockedDownloadSnapshot,
  } as unknown as Snapshot);

  mockedSnapshotFactory.createFromExistingSnapshot.mockResolvedValue({
    delete: mockedDeleteSnapshot,
    download: mockedDownloadSnapshot,
  } as unknown as Snapshot);
};

describe('org:resources:pull', () => {
  beforeAll(() => {
    doMockConfig();
    mockAuthenticatedClient();

    mockedGetSnapshot.mockResolvedValue(
      getDummySnapshotModel('default-org', 'my-snapshot', [
        getSuccessReport('my-snapshot', ResourceSnapshotsReportType.Apply),
      ])
    );
  });

  beforeEach(() => {
    doMockSnapshotFactory();
    doMockPreconditions();
    mockUserHavingAllRequiredPlatformPrivileges();
  });

  afterEach(() => {
    mockEvaluate.mockClear();
    mockedProject.mockClear();
    mockedIsGitInstalled.mockClear();
    mockedSnapshotFactory.mockClear();
  });

  test
    .do(() => {
      mockUserNotHavingAllRequiredPlatformPrivileges();
    })
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .catch(/You are not authorized to create snapshot/)
    .it('should return an error message if platform privileges are missing');

  test
    .do(() => {
      mockUserNotHavingAllRequiredResourcePrivileges();
    })
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .catch((ctx) => {
      expect(ctx.message).toMatchSnapshot();
    })
    .it('should return an error message if resource privileges are missing');

  test
    .do(() => {
      mockUserNotHavingAllRequiredResourcePrivileges();
    })
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-s', 'some-snapshot-id'])
    .catch((ctx) => {
      expect(ctx.message).toMatchSnapshot();
    })
    .it(
      'should return an error message if resource privileges are missing for specific snapshot'
    );

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should use the cwd as project', () => {
      expect(mockedProject).toHaveBeenCalledWith(cwd(), expect.anything());
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-o', 'someorgid'])
    .it('should create Project with someOrgId', () => {
      expect(mockedProject).toHaveBeenCalledWith(
        expect.anything(),
        'someorgid'
      );
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should download the snapshot', () => {
      expect(mockedDownloadSnapshot).toHaveBeenCalled();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should write the project resources manifest', () => {
      expect(mockedProject.prototype.writeResourcesManifest).toHaveBeenCalled();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should delete the snapshot', () => {
      expect(mockedDeleteSnapshot).toHaveBeenCalled();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-s', 'someSnapshotId'])
    .it('should not delete the snapshot', () => {
      expect(mockedDeleteSnapshot).not.toHaveBeenCalled();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should select all resource types', () => {
      const resourcesToExport = {
        FIELD: ['*'],
        FEATURED_RESULT: ['*'],
        SOURCE: ['*'],
        QUERY_PIPELINE: ['*'],
        SEARCH_PAGE: ['*'],
        EXTENSION: ['*'],
      };
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        expect.objectContaining(resourcesToExport),
        'default-org',
        expect.objectContaining({})
      );
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-r', 'FIELD', 'FEATURED_RESULT', 'SOURCE'])
    .it('should select specified resource types', () => {
      const resourcesToExport = {
        FIELD: ['*'],
        FEATURED_RESULT: ['*'],
        SOURCE: ['*'],
      };
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        resourcesToExport,
        'default-org',
        expect.objectContaining({})
      );
    });

  test
    .stdout()
    .stderr()
    .stub(CliUx.ux, 'confirm', () => () => Promise.resolve(true))
    .command([
      'org:resources:pull',
      '-m',
      join(pathToStub, 'snapshotPullModels', 'full.json'),
    ])
    .it(
      'should create a snapshot with all resource types specified in the model',
      () => {
        const resourcesToExport = {
          FIELD: ['*'],
          QUERY_PIPELINE: ['*'],
          ML_MODEL: ['*'],
          SUBSCRIPTION: ['*'],
        };
        expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
          resourcesToExport,
          expect.anything(),
          expect.objectContaining({})
        );
      }
    );

  test
    .stdout()
    .stderr()
    .stub(CliUx.ux, 'confirm', () => () => Promise.resolve(true))
    .command([
      'org:resources:pull',
      '-m',
      join(pathToStub, 'snapshotPullModels', 'subset.json'),
    ])
    .it(
      'should create a snapshot with only the specified resource items in the model',
      () => {
        const resourcesToExport = {
          FIELD: ['author', 'source', 'title'],
          QUERY_PIPELINE: ['default', 'agentPanel'],
          ML_MODEL: null,
          SEARCH_PAGE: null,
          SUBSCRIPTION: null,
        };
        expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
          resourcesToExport,
          expect.anything(),
          expect.objectContaining({})
        );
      }
    );

  test
    .stdout()
    .stderr()
    .stub(CliUx.ux, 'confirm', () => () => Promise.resolve(true))
    .command([
      'org:resources:pull',
      '-m',
      join(pathToStub, 'snapshotPullModels', 'full.json'),
    ])
    .it('should use the orgId from the model', () => {
      expect(mockedProject).toHaveBeenCalledWith(expect.anything(), 'myorgid');
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        expect.anything(),
        'myorgid',
        expect.objectContaining({})
      );
    });

  test
    .stdout()
    .stderr()
    .command([
      'org:resources:pull',
      '-m',
      join(pathToStub, 'snapshotPullModels', 'missingOrgId.json'),
    ])
    .it('should use the orgId from the config', () => {
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        expect.anything(),
        'default-org',
        expect.objectContaining({})
      );
    });

  test
    .stdout()
    .stderr()
    .command([
      'org:resources:pull',
      '-m',
      join(pathToStub, 'snapshotPullModels', 'missingResources.json'),
    ])
    .catch(/requires property "resourcesToExport"/)
    .it('should throw an invalid model error');

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull'])
    .it('should set a 60 seconds timeout', () => {
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        expect.arrayContaining([]),
        'default-org',
        {wait: 60}
      );
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-w', '78'])
    .it('should set a 78 seconds timeout', () => {
      expect(mockedSnapshotFactory.createFromOrg).toHaveBeenCalledWith(
        expect.arrayContaining([]),
        'default-org',
        {wait: 78}
      );
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:pull', '-r', 'invalidresource'])
    .catch(/Expected --resourceTypes=invalidresource to be one of/)
    .it('should not allow invalid resource');
});
