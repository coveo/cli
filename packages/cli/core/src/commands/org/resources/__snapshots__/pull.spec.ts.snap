// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org:resources:pull should return an error message if resource privileges are missing 1`] = `
"You are missing [1mREAD[22m privilege on the following resources
 â€¢ SOURCE
 â€¢ ML_MODEL

Visit https://docs.coveo.com/en/3357 for more info on privileges required to manage snapshots."
`;

exports[`org:resources:pull should return an error message if resource privileges are missing for specific snapshot 1`] = `
"You do not have [1mREAD[22m privilege on the snapshot some-snapshot-id
Visit https://docs.coveo.com/en/3357 for more info on privileges required to manage snapshots."
`;
